### ISSUE TITLE
Directly edit source code

### ISSUE THREAD

**Issue by bernhardh** [heyscrumpy/tiptap#58]
Is there a way, to directly edit html source code? I mean, is there a way, to add a button to toggle wysiwyg view to source-code view?

**Comment by philippkuehn** [comment]
This should be possible right now. There is an `@update` event on the editor component where you can grab the current HTML. Then you could add a toggle button and render this HTML in a textarea. And on change you could set the new content with `this.$refs.editor.setContent('<div>your html</div>')`.

Would be a nice example for the demo page btw.

**Comment by bilouw** [comment]
Hi,

I need to render in some case the html content directly in the editor. If copy paste, that works, but when i set the content on editor creation, i can't figure out how to render html source code.

> Then you could add a toggle button and render this HTML in a textarea.

Are you talking about another text area than the editor ? I don't understand how can i do what you said

Thanks.

**Comment by silvaDominic** [comment]
> EDIT: I just saw that a feature has been released to view source code. How can i use this feature ?

@bilouw Did you ever find out how to use this feature? I'm scouring the docs and google and can't find anything. I prefer not to do the hack mentioned above.

**Comment by bilouw** [comment]
@silvaDominic Nope, i just used some tricks to escape or unescape the html. This way, i can render html directly in the editor.

**Comment by josuermartinezm** [comment]
@bilouw how did it?

**Comment by digthedill** [comment]
I fourth this notion. My solutions are too hacky

**Comment by ITwrx** [comment]
Yes, using the editing menus/buttons is best for users most of the time, but sometimes people need to edit the source code. Also, more high-level/legacy/typical editor libraries have already had this feature for years so, if it's doable with tiptap in a clean, supported way, it should be included in the Examples docs, IMHO. 

IIRC, some of these other editors don't have syntax highlighting on the viewable/editable source though, so it would be nice if the example showed how to integrate that, as well.

**Comment by acoyfellow** [comment]
I won't be able to use Tiptap if I can't figure out a solution to editing source. 

+1 to adding an example to docs.

**Comment by ringge** [comment]
Did anyone figure out a clean way to edit HTML source in Tiptap?

**Comment by AhmedAlsumairi** [comment]
Any Update regrading to edit source Code

**Comment by RadekHavelka** [comment]
you just take editor.getHTML() and put it into textarea.value, then, after user is done with editing, you take textarea.value and use editor.setContent() with that value ...

```
    editor.commands.clearContent(true);
    editor.commands.setContent(textrea.value, true);
```

**Comment by Maneesh43** [comment]
I did this way,
To show source:
```
editor.clearContent()
edtor.setContent(`<textarea>${editor.getHTML()}</textarea>`);
```

to Toggle back:
`editor.setContent(editor.getText())`
for some reason now editor.getHTML() returns escaped html so just using getText()

**Comment by OliverGrimsley** [comment]
For anyone that gets here, using vue (with Ref value) - and slight update to @Maneesh43 I added 2 buttons to show and hide HTML source - did not bother with a boolean, and these two functions (note that `clearContent()` first did not work for me, I needed backticks inside the parens, and I think they updated the code to make these functions under `commands`):

```javascript
const showSource = () => {
  editor.value.commands.setContent(`<textarea>${editor.value.getHTML()}</textarea>`)
}

const showHTML = () => {
  editor.value.commands.setContent(editor.value.getText())
}
```

**Comment by muhammetakkus** [comment]
How can I display the HTML in a cleaner, CKEditor-like format, where there’s at least one line break between tags? Also, when I add attributes to tags, they aren’t preserved — the editor removes them when switching views. For example, I want to be able to write and apply something like <p--->style="color:red"> manually.
